internal/core/component_factory.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+#![warn(missing_docs)]
+
+//! This module contains the `ComponentFactory` that is exposed to the C API
+
+use vtable::*;
+
+use crate::component::ComponentRc;
+
+/// A `ComponentFactory` used can create a `Component`
+#[vtable]
+#[repr(C)]
+pub struct ComponentFactoryVTable {
+    /// Visit the children of the item at index `index`.
+    /// Note that the root item is at index 0, so passing 0 would visit the item under root (the children of root).
+    /// If you want to visit the root item, you need to pass -1 as an index.
+    pub build: extern "C" fn(core::pin::Pin<VRef<ComponentFactoryVTable>>) -> Option<ComponentRc>,
+
+    /// in-place destructor (for VRc)
+    pub drop_in_place: unsafe fn(VRefMut<ComponentFactoryVTable>) -> vtable::Layout,
+    /// dealloc function (for VRc)
+    pub dealloc: unsafe fn(&ComponentFactoryVTable, ptr: *mut u8, layout: vtable::Layout),
+}
+
+/// A implementation of a `vtable::VRc` for `ComponentFactoryVTable`
+pub type ComponentFactoryRc = vtable::VRc<ComponentFactoryVTable>;
+
+/// A Box containing a `ComponentFactory` or nothing
+#[derive(Default, Debug)]
+pub struct ComponentFactoryBox(Option<ComponentFactoryRc>);
```

Comment:
I think we want to have something like
```rust
struct ComponentFactoryRc(Option<vtable::VRc<ComponentFactoryVTable>>>);
``` 

This is somehow matching what ModelRc is.

- No need for a type alias (your current `ComponentFactoryRc`  and also no need for `ComponentFactoryBox` which is not a box